// verdaccio/build/lib/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "API_ERROR", {
  enumerable: true,
  get: function () {
    return _core.API_ERROR;
  }
});
Object.defineProperty(exports, "API_MESSAGE", {
  enumerable: true,
  get: function () {
    return _core.API_MESSAGE;
  }
});
Object.defineProperty(exports, "APP_ERROR", {
  enumerable: true,
  get: function () {
    return _core.APP_ERROR;
  }
});
Object.defineProperty(exports, "CHARACTER_ENCODING", {
  enumerable: true,
  get: function () {
    return _core.CHARACTER_ENCODING;
  }
});
exports.ERROR_CODE = exports.DIST_TAGS = exports.DEFAULT_USER = exports.DEFAULT_UPLINK = exports.DEFAULT_REGISTRY = exports.DEFAULT_PROTOCOL = exports.DEFAULT_PORT = exports.DEFAULT_NO_README = exports.DEFAULT_MIN_LIMIT_PASSWORD = exports.DEFAULT_DOMAIN = void 0;
Object.defineProperty(exports, "HEADERS", {
  enumerable: true,
  get: function () {
    return _core.HEADERS;
  }
});
Object.defineProperty(exports, "HEADER_TYPE", {
  enumerable: true,
  get: function () {
    return _core.HEADER_TYPE;
  }
});
Object.defineProperty(exports, "HTTP_STATUS", {
  enumerable: true,
  get: function () {
    return _core.HTTP_STATUS;
  }
});
exports.MODULE_NOT_FOUND = exports.LATEST = void 0;
Object.defineProperty(exports, "PACKAGE_ACCESS", {
  enumerable: true,
  get: function () {
    return _utils.PACKAGE_ACCESS;
  }
});
Object.defineProperty(exports, "ROLES", {
  enumerable: true,
  get: function () {
    return _utils.ROLES;
  }
});
exports.STORAGE = void 0;
Object.defineProperty(exports, "SUPPORT_ERRORS", {
  enumerable: true,
  get: function () {
    return _core.SUPPORT_ERRORS;
  }
});
exports.TIME_EXPIRATION_24H = exports.TIME_EXPIRATION_1H = void 0;
Object.defineProperty(exports, "TOKEN_BASIC", {
  enumerable: true,
  get: function () {
    return _core.TOKEN_BASIC;
  }
});
Object.defineProperty(exports, "TOKEN_BEARER", {
  enumerable: true,
  get: function () {
    return _core.TOKEN_BEARER;
  }
});
exports.keyPem = exports.csrPem = exports.certPem = exports.WEB_TITLE = exports.USERS = void 0;
var _core = require("@verdaccio/core");
var _utils = require("@verdaccio/utils");
const DEFAULT_PORT = '4873';
exports.DEFAULT_PORT = DEFAULT_PORT;
const DEFAULT_PROTOCOL = 'http';
exports.DEFAULT_PROTOCOL = DEFAULT_PROTOCOL;
const DEFAULT_DOMAIN = 'localhost';
exports.DEFAULT_DOMAIN = DEFAULT_DOMAIN;
const TIME_EXPIRATION_24H = '24h';
exports.TIME_EXPIRATION_24H = TIME_EXPIRATION_24H;
const TIME_EXPIRATION_1H = '1h';
exports.TIME_EXPIRATION_1H = TIME_EXPIRATION_1H;
const DIST_TAGS = 'dist-tags';
exports.DIST_TAGS = DIST_TAGS;
const LATEST = 'latest';
exports.LATEST = LATEST;
const USERS = 'users';
exports.USERS = USERS;
const DEFAULT_MIN_LIMIT_PASSWORD = 3;
exports.DEFAULT_MIN_LIMIT_PASSWORD = DEFAULT_MIN_LIMIT_PASSWORD;
const DEFAULT_USER = 'Anonymous';
exports.DEFAULT_USER = DEFAULT_USER;
const keyPem = 'verdaccio-key.pem';
exports.keyPem = keyPem;
const certPem = 'verdaccio-cert.pem';
exports.certPem = certPem;
const csrPem = 'verdaccio-csr.pem';
exports.csrPem = csrPem;
const ERROR_CODE = {
  token_required: 'token is required'
};
exports.ERROR_CODE = ERROR_CODE;
const DEFAULT_REGISTRY = 'https://registry.npmjs.org';
exports.DEFAULT_REGISTRY = DEFAULT_REGISTRY;
const DEFAULT_UPLINK = 'npmjs';
exports.DEFAULT_UPLINK = DEFAULT_UPLINK;
const DEFAULT_NO_README = 'ERROR: No README data found!';
exports.DEFAULT_NO_README = DEFAULT_NO_README;
const MODULE_NOT_FOUND = 'MODULE_NOT_FOUND';
exports.MODULE_NOT_FOUND = MODULE_NOT_FOUND;
const WEB_TITLE = 'Verdaccio';
exports.WEB_TITLE = WEB_TITLE;
const STORAGE = {
  PACKAGE_FILE_NAME: 'package.json',
  FILE_EXIST_ERROR: 'EEXISTS',
  NO_SUCH_FILE_ERROR: 'ENOENT',
  DEFAULT_REVISION: '0-0000000000000000'
};
exports.STORAGE = STORAGE;
